Implements transit transport

pfeodrippe opened the pull request:
### When and if ready, the code to implement the transit transport will be in some other library as said at #61.

Before submitting a PR make sure the following things have been done:

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests to cover your change(s)
- [x] All tests are passing
- [x] The new code is not generating reflection warnings
- [x] You've updated the [changelog](../CHANGELOG.md)(that only applies to user-visible changes)

Thanks!

pfeodrippe commented:
I'm in doubt if message pack is the best way type to go on, but json didn't work very well, I still have to check why, if someone could give me a hint...

bbatsov commented:
Excellent work! 🙇 

Seems you also tackled some of the outstanding issues for #67, so I guess you can collaborate with @dotemacs to wrap the other PR together. 

> I'm in doubt if message pack is the best type to go on, but json didn't work very well, I still have to check why, does anyone has any idea?

I think that ideally the transit transport should support both messagepack and json (basically we'd have two transports that users would be able to choose from). What were the problems that you encountered with JSON?

bbatsov commented on src/clojure/nrepl/transport.clj:
```diff
@@ -79,7 +80,7 @@
        (throw (SocketException. "The transport's socket appears to have lost its connection to the nREPL server")))
      (catch Throwable e#
-       (if (and ~s (not (.isConnected ~s)))
+       (if (and ~s (not (.isClosed ~s)))
```
I don't understand the need for this change. What's the motivation for it?

pfeodrippe replied:
At `nrepl.core-test`, the tests `clients-fail-on-disconnects` and `transports-fail-on-disconnects` failswhen using `.isConnected` gives me an error, both using the transit transport with messagepack.

pfeodrippe replied:
Sorry, after changing the implementation, I saw that it was not affecting the tests, so it will be the same as before

dotemacs commented:
👍

pfeodrippe commented:
Taken some code from your #67, @dotemacs, thanks =)!

> I think that ideally the transit transport should support both messagepack and json (basically we'd have two transports that users would be able to choose from). What were the problems that you encountered with JSON?

Was having some parser problems using `json`, maybe I need to understand more about transit (never used it anywhere) to make this type work with nrepl, I think I need some help here. If someone would have time to test, just go to `nrepl.transport` and change 

```clojure
(transit/read
 (transit/reader in :msgpack))
```
to 
```clojure
(transit/read
 (transit/reader in :json))
```
and
```clojure
(transit/write
 (transit/writer out :msgpack)
 %)
```
```clojure
(transit/write
 (transit/writer out :json)
 %)
```

pfeodrippe commented:
Now `transit+msgpack`, `transit+json` and `transit+json-verbose` are supported (at least regarding the tests)

bbatsov commented:
> Now transit+msgpack, transit+json and transit+json-verbose are supported (at least regarding the tests)

Excellent!

Did you also test the transports manually to make sure that they behave properly?

pfeodrippe commented:
> Did you also test the transports manually to make sure that they behave properly?

Hope to do some manual tests soon, now I'm trying to tackle at #50, it's very annoying for multiple transports. Transports `bencode` and `transit+msgpack` are failing at this test... sometimes, while `transit+json` and `transit+json-verbose` are not failing anytime, very strange!!

What I found now is that for `bencode` and `transit+msgpack`, using some `log/info` for debug, the state of the server after `(.close server)` at `transports-fail-on-disconnects` is false inside the write function, while it's true for `transit+json` and `transit+json-verbose`. Gonna follow this lead a little more today.

bbatsov commented:
Great! This was a major annoyance for me as well!

On Sun, 18 Nov 2018 at 19:07, Paulo Rafael Feodrippe <
notifications@github.com> wrote:

> Did you also test the transports manually to make sure that they behave
> properly?
>
> Hope to do some manual tests soon, now I'm trying to tackle at #50
> <https://github.com/nrepl/nrepl/issues/50>, it's very annoying for
> multiple transports. Transports bencode and transit+msgpack are failing
> at this test... sometimes, while transit+json and transit+json-verbose
> are not failing anytime, very strange!!
>
> What I found now is that for bencode and transit+msgpack, using some
> log/info for debug, the state of the server after (.close server) at
> transports-fail-on-disconnects is false inside the write function, while
> it's true for transit+json and transit+json-verbose. Gonna follow this
> lead a little more today.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nrepl/nrepl/pull/84#issuecomment-439712849>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGVyn9M86u72ttAgtDSuMcjJRXd_VE6ks5uwaHkgaJpZM4YnpWN>
> .
>
-- 
Best Regards,
Bozhidar Batsov

http://www.batsov.com

bbatsov commented:
Btw, I've also created an empty project for the transports https://github.com/nrepl/fastlane

Once we feel that they are ready we'll release them as `nrepl/fastlane`. I decided that `nrepl/transit` wasn't a proper nREPL project name. :D I'm not as poetic as Chas, but I do my best! 😆

pfeodrippe commented:
Great! Hope to fill it soon with not so poetic, but useful, code \o/

pfeodrippe commented:
There are still things to do (tests, cmdline verbose support), but I want to hear what you have to say before going further

bbatsov commented:
@pfeodrippe I like the approach!

I'm thinking that maybe for the function `logger` might be a better name than `default-verbose`/`verbose-fn`, but overall everything seems solid.

pfeodrippe commented:
> I'm thinking that maybe for the function `logger` might be a better name than `default-verbose`/`verbose-fn`, but overall everything seems solid.

Thanks! Gonna change it.

> Btw, did you add the logging to this PR accidentally? I think it's going to be better to just handle this in a separate branch/PR.

I had to test the different transports communication data with a new test based at the logger as you said at #85 that I would needed.  As they are orthogonal to each other (one is at `transport.clj` and the other at `server.clj`), it would be easy to split and make a new PR for #85 with this new test.

pfeodrippe commented:
What would be put at `nrepl/fast-lane`? 

I see that `nrepl/bencode` is needed because the protocol is implemented there, but what are the benefits for transit?

bbatsov commented:
> I see that nrepl/bencode is needed because the protocol is implemented there, but what are the benefits for transit?

We can't have an explicit dep to transit in nREPL that's why the transport should go to the other repo. For bencode is not dangerous, as this was extracted from nREPL and chance of someone accidentally running into dep conflicts is very small. Foundational libs like nREPL should ideally have 0 runtime deps themselves, as they are running in the context of the user app. We can obfuscate the transit dep, but I prefer it to just be a separate project like drawbridge. Tiny and focused.

pfeodrippe commented:
Removed code referring to `transit`, added a profile to test it with all `nrepl.core-test` code and added it to `test-all` lein alias, so it can be tested as well as `bencode` and other future internal transports.

bbatsov commented on test/clojure/nrepl/core_test.clj:
```diff
@@ -259,6 +293,7 @@
     (is (= [true] (repl-values session "(.contains (str *e) \"bad, bad code\")")))))
 
 (def-repl-test multiple-expressions-return
+  (session {:op :bypass :stdin "bypassing..."})
```
Btw, what's this? Seems a bit random.

pfeodrippe replied:
Sorry, was just some debugging

bbatsov commented on src/clojure/nrepl/server.clj:
```diff
@@ -2,6 +2,7 @@
   "Default server implementations"
   {:author "Chas Emerick"}
   (:require
+   [clojure.java.io :as io]
```
I think that's not needed.

bbatsov commented on CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@
 * [#68](https://github.com/nrepl/nREPL/issues/68): Avoid illegal access warning on JDK 9+ caused by `nrepl.middleware.interruptible-eval/set-line!`.
 * [#77](https://github.com/nrepl/nREPL/issues/77): Exit cleanly after pressing `ctrl-d` in an interactive REPL.
 * [#13](https://github.com/nrepl/nREPL/issues/13): Catch ThreadDeath exception thrown by interrupt.
+* [#61](https://github.com/nrepl/nrepl/issues/61): Implement a Transit transport
```
This should be reflected in the manual, not in the changelog (as it's a different project).

pfeodrippe replied:
You mean update the docs at https://nrepl.xyz/nrepl/design/transports.html?

bbatsov replied:
Or the list of third-party extensions.

pfeodrippe replied:
Added fastlane to https://github.com/nrepl/nrepl/wiki/Extensions

bbatsov commented:
You'll also have to rebase on top of `master` it seems.

pfeodrippe commented:
Added fastlane to https://github.com/nrepl/nrepl/wiki/Extensions

pfeodrippe commented:
Also rebased the code to `master`.

bbatsov commented:
Thanks!

pfeodrippe commented:
Thank you o/
